DuckDuckGo/AutofillPixelReporter.swift

```diff
@@ -0,0 +1,149 @@
+
+final class AutofillPixelReporter {
+
+    @UserDefaultsWrapper(key: .autofillSearchDauDate, defaultValue: .distantPast)
+    var autofillSearchDauDate: Date
+
+    @UserDefaultsWrapper(key: .autofillFillDate, defaultValue: .distantPast)
+    var autofillFillDate: Date
+
+    @UserDefaultsWrapper(key: .autofillOnboardedUser, defaultValue: false)
+    var autofillOnboardedUser: Bool
+
+    private let statisticsStorage: StatisticsStore
+    private let secureVault: (any AutofillSecureVault)?
+
+    enum EventType {
+        case fill
+        case searchDAU
+    }
+
+    init(statisticsStorage: StatisticsStore = StatisticsUserDefaults(), secureVault: (any AutofillSecureVault)? = try? AutofillSecureVaultFactory.makeVault(reporter: SecureVaultReporter.shared)) {
+        self.statisticsStorage = statisticsStorage
+        self.secureVault = secureVault
+
+        createNotificationObservers()
+
+        if shouldFireOnboardedUserPixel() {
+            DailyPixel.fire(pixel: .autofillOnboardedUser)
+        }
+    }
+
+    private func createNotificationObservers() {
+        NotificationCenter.default.addObserver(self, selector: #selector(didReceiveSearchDAU), name: .searchDAU, object: nil)
+        NotificationCenter.default.addObserver(self, selector: #selector(didReceiveFillEvent), name: .autofillFillEvent, object: nil)
+    }
+
+    @objc
+    private func didReceiveSearchDAU() {
+        guard !Date().isSameDay(autofillSearchDauDate) else {
+            return
+        }
+
+        autofillSearchDauDate = Date()
+
+        firePixels(pixelsToFireFor(.searchDAU))
+    }
+
+    @objc
+    private func didReceiveFillEvent() {
+        guard !Date().isSameDay(autofillFillDate) else {
+            return
+        }
+
+        autofillFillDate = Date()
+
+        firePixels(pixelsToFireFor(.fill))
+    }
+
+    func pixelsToFireFor(_ type: EventType) -> [Pixel.Event] {
+        var pixelsToFire: [Pixel.Event] = []
+
+        if shouldFireActiveUserPixel() {
+            pixelsToFire.append(.autofillActiveUser)
+        }
+
+        switch type {
+        case .fill:
+            pixelsToFire.append(.autofillLoginsStacked)
+        case .searchDAU:
+            if shouldFireEnabledUserPixel() {
+                pixelsToFire.append(.autofillEnabledUser)
+            }
+        }
+
+        return pixelsToFire
+    }
+
+    private func firePixels(_ pixels: [Pixel.Event]) {
+        for pixel in pixels {
+            switch pixel {
+            case .autofillLoginsStacked:
+                let bucket = (try? secureVault?.accountsCountBucket()) ?? ""
+                DailyPixel.fire(pixel: pixel, withAdditionalParameters: [PixelParameters.countBucket: bucket])
+            default:
+                DailyPixel.fire(pixel: pixel)
+            }
+        }
+    }
+
+    private func shouldFireActiveUserPixel() -> Bool {
+        let today = Date()
+        if today.isSameDay(autofillSearchDauDate) && today.isSameDay(autofillFillDate) {
```

Comment:
If you wanted you could change this to:

```suggestion
        if Date().isSameDay(autofillSearchDauDate) && today.isSameDay(autofillFillDate) {
```

…just to be consistent with how you use `Date()` in the other functions above and below

Reply:
ah, that was actually intentional as I'm referencing `today` twice in that if condition 🙂